### ISSUE TITLE
Escaped dollar signs in heredoc blocks are not recognized correctly.

### ISSUE THREAD

**Issue by lucassus** [hadolint/hadolint#1011]
- [x] This is a bug report
- [ ] This is a feature request
- [ ] I searched existing issues before opening this one

I have the following RUN with heredoc in my Dockerfile

```Dockerfile
ARG WKHTMLTOPDF_VERSION="0.12.6.1-3"

# Install patched wkhtmltopdf for PDF generation, see https://github.com/wkhtmltopdf/packaging/releases
RUN <<EOF bash
  set -eux

  # We need to escape dollar sign in variables to prevent it from being expanded during the build.
  # For instance use \${version_codename}.
  version_codename=$(grep --only-matching --perl-regexp '(?<=VERSION_CODENAME=).+$' /etc/os-release)
  download_url="https://github.com/wkhtmltopdf/packaging/releases/download/${WKHTMLTOPDF_VERSION}/wkhtmltox_${WKHTMLTOPDF_VERSION}.\${version_codename}_${BUILDARCH}.deb"

  curl \${download_url} --location --output /tmp/wkhtmltopdf.deb
  apt-get install --yes --no-install-recommends /tmp/wkhtmltopdf.deb
  rm /tmp/wkhtmltopdf.deb
EOF
```

It work fine. But the linter complains with:

```
-:48 SC2034 warning: download_url appears unused. Verify use (or export if used externally).
```

### Expected behavior

The linter should understand that I have to escape a dollar sign in this case.

**Comment by m-ildefons** [comment]
I'm not sure there really is a bug here.

I don't see a reason why the `$` needs to be escaped in the heredoc.
With this minimal reproducer dockerfile:
```dockerfile
FROM debian:bullseye

ARG CANARY

RUN <<EOF
#!/bin/bash -eux
foobar="The canary makes ${CANARY}"
echo "${foobar}" >> logfile
EOF

RUN cat logfile
```
and the build command `docker buildx build --build-arg CANARY="chirp chirp" -t test . --no-cache  --progress plain`, I get the following output:

<img width="1808" height="541" alt="Image" src="https://github.com/user-attachments/assets/949e3844-c126-46f1-93a2-5704db13e3e6" />

As you can see from the third to last line of the output, the build argument and the bash variable were both expanded correctly within their respective contexts without the need to escape the `$`.
This dockerfile also does not produce either SC1083 or SC2034 as warnings with Hadolint v2.14.0.

If I do escape the `$`, then bash no longer expands the variable `foobar`:

<img width="1781" height="556" alt="Image" src="https://github.com/user-attachments/assets/3f756ad0-07ac-42db-a00c-64fc2ac85999" />

However in this case Hadolint correctly emits SC2034, because the variable `foobar` really is no longer used.

If I don't quote the variable expansion of `foobar` and I also escape the `$`, then I get both errors SC1083 and SC2034:

<img width="1802" height="684" alt="Image" src="https://github.com/user-attachments/assets/c9979c25-af59-4ed4-b207-b754de88a864" />

but that is correct again, because the variable `foobar` really isn't referenced either and the  `}` character is literal. In this particular case, this isn't a bug per sé, but the warning SC1083 is still correctly emitted, because the code isn't following best practices around quoting literal data.
